### PR TITLE
Enable eager checks for memory sanitizer

### LINKED ISSUE #99179: Enable eager checks for memory sanitizer
The memory sanitizer supports "eager checks", which will check initialization for noundef parameters and return values. It can be enabled by passing true as the fourth parameter to MemorySanitizerOptions. The corresponding clang option is `-fsanitize-memory-param-retval`. Not sure whether we need an option or can just unconditionally enable it.

### PR DESCRIPTION
**Pull request by ghost**
Fixes #99179

### CODE REVIEW

**Comment by 5225225** [comment]
Question: @nikic , were the eager checks added in LLVM 15? (So I'd add an `#if LLVM_VERSION_GE(14, 0)` and only pass the eager check if then?). Or was it an earlier version? I can't seem to find historical llvm docs, and https://llvm.org/doxygen/structllvm_1_1MemorySanitizerOptions.html doesn't mention the added version, so I don't know where I'd go to find out.

**Comment by nikic** [comment]
The option was added in LLVM 14.

**Comment by 5225225** [comment]
Strange, I was expecting that run to fail because LLVM version too old, and I'd need to put a min-llvm-version on memory-eager.rs

File: compiler/rustc_codegen_llvm/src/abi.rs [hunk 1]
```diff
@@ -90,6 +91,13 @@ fn get_attrs<'ll>(this: &ArgAttributes, cx: &CodegenCx<'ll, '_>) -> SmallVec<[&'
         if regular.contains(ArgAttribute::NoAliasMutRef) && should_use_mutable_noalias(cx) {
             attrs.push(llvm::AttributeKind::NoAlias.create_attr(cx.llcx));
         }
+    } else if cx.tcx.sess.opts.debugging_opts.sanitizer.contains(SanitizerSet::MEMORY) {
```
[the remark this answers is not in the document]
Comment (5225225) [reply compiler/rustc_codegen_llvm/src/abi.rs]: also came across a use of debugging_opts in the rustc dev guide when grepping for any places i missed, fixed that too (on their repo) :)

File: src/test/ui/sanitize/memory.rs [hunk 2]
```diff
@@ -1,12 +1,16 @@
...
+//
+// [optimized]compile-flags: -Z sanitizer=memory -Zsanitizer-memory-track-origins -O
+// [unoptimized]compile-flags: -Z sanitizer=memory -Zsanitizer-memory-track-origins
 //
 // run-fail
 // error-pattern: MemorySanitizer: use-of-uninitialized-value
 // error-pattern: Uninitialized value was created by an allocation
-// error-pattern: in the stack frame of function 'main'
+// [optimized]error-pattern: in the stack frame of function 'main'
+// [unoptimized]error-pattern: in the stack frame of function 'random'
```
[the remark this answers is not in the document]
Comment (5225225) [reply src/test/ui/sanitize/memory.rs]: seems to, yeah. thanks!

**Comment by tmiasko** [comment]
From the perspective of the Rust alone enabling eager checks by default sounds fine to me. 

Though I wonder about mixed C/C++ builds. Clang disables eager checks by default, it also has a different understanding of undef types. Both cases could lead to false negatives. The side with eager checks might falsely consider the caller to be responsible for checking the undef argument and assumes the argument to be initialized.

For that reason I would lean towards exposing this functionality behind a flag and matching the clang default. What do you think?

**Comment by 5225225** [comment]
A LLVM person would need to answer if that's a problem.

But I do know that you get false *positives* if you don't have all code built with memory sanitizer, so "being careful about your dependency build flags" is already a thing.

I'm wary of making it opt-in since *most* code doesn't need it to be opt in.

We could add a mandatory parameter for eager checks and show a warning if the user doesn't pick one, perhaps. That'll communicate this change pretty clearly, but is rather loud.

Not sure.

**Comment by tmiasko** [comment]
> A LLVM person would need to answer if that's a problem.

It is a problem. It is impossible to detect uninitialized arguments of integer types, floating point types, and data pointers in calls from Rust to C when eager checks enabled in both clang and rustc. Currently the Rust side doesn't have undef attribute for those types, so argument is not checked. The C side does have the attribute and assumes it was already checked.

There is an analogous issue when the eager checks are enabled only in rustc.

The implementation that preserves the existing default sound fine to me, if you would like to land it. With regards to the change of the default, I would wait for further support of the idea.

**Comment by cuviper** [comment]
The code looks fine, but I don't have enough background in msan to judge if we *should*, so I'm re-rolling...

r? rust-lang/compiler

**Comment by jackh726** [comment]
I have less background to know if we should or not.

Let's cc @rust-lang/compiler

**Comment by jackh726** [comment]
I don't really have any background here to judge if this is a good change or not. I imagine this might need a compiler MCP, but not sure. Going to go ahead and nominate to get discussed in the compiler meeting, since that seems like to easiest way to get feedback here.

**Comment by jackh726** [comment]
Talked about this in [compiler meeting](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202022-09-08/near/297800349). Seems like we're good to just enable this.

@bors r+

**Pull request merged**